app/jobs/maintenance_tasks/task_job.rb

```diff
@@ -25,21 +25,14 @@ def retry_on(*, **)
 
     private
 
+    EnumerationContext = Struct.new(:cursor, keyword_init: true)
+
     def build_enumerator(_run, cursor:)
-      cursor ||= @run.cursor
-      collection = @task.collection
-
-      case collection
-      when ActiveRecord::Relation
-        enumerator_builder.active_record_on_records(collection, cursor: cursor)
-      when Array
-        enumerator_builder.build_array_enumerator(collection, cursor: cursor)
-      when CSV
-        JobIteration::CsvEnumerator.new(collection).rows(cursor: cursor)
-      else
-        raise ArgumentError, "#{@task.class.name}#collection must be either "\
-          'an Active Record Relation, Array, or CSV.'
-      end
+      context = EnumerationContext.new(
+        cursor: cursor || @run.cursor,
+      )
+
+      @task.enumerator_builder.enumerator(context: context)
```

Comment:
We build a context object instead of coupling to a `cursor:` keyword argument to allow forward compatibility, should we need to pass in more in the future, as @rafaelfranca mentioned in https://github.com/Shopify/maintenance_tasks/pull/310#issuecomment-767783835.

Reply:
Yet we're now coupling it to the `context:` keyword argument. I think `enumerator`should have an normal parameter.

Reply:
The reasoning here wasn't that we didn't want to couple to a keyword argument, it was that if we ever needed to add more to the context, we'd have to make a breaking change to the signature.


```ruby
class SomeConsumersEnumeratorBuilder
  def enumerator(cursor:)
  end
end

@task.enumerator_builder.enumerator(cursor: cursor, new_thing: 123) # 💥 ArgumentError
```

```ruby
class SomeConsumersEnumeratorBuilder
  def enumerator(cursor)
  end
end

@task.enumerator_builder.enumerator(cursor, 123) # 💥 ArgumentError
```

By instead passing in a context object, we can simply add methods to it, and if the consumer doesn't care, they don't call it; no breaking change.

```ruby
class SomeConsumersEnumeratorBuilder
  def enumerator(context:)
  end
end

@task.enumerator_builder.enumerator(context: fancy_new_context) # ✅
```

As for the choice of a keyword argument, I've found it clarifies what to expect when it isn't obvious from the method name (compared to something like `delete_user(user)`). Here it makes it entirely unambiguous to consumers that they're receiving a context object, not a cursor or something else.

I can certainly change it to a positional argument if you think that's more appropriate, but that's why I chose `context:`.

Reply:
Also +1 to Sam's comments here, I think foregoing the keyword argument results in too much ambiguity. Especially with users who are coming from having used JobIteration, I think it's highly likely they'd write `def enumerator(cursor)` and expect their method to receive a cursor object, since JobIteration's API was `build_enumerator(cursor:)`

I also think coupling it to something as generic as `context` gives us the flexibility to expand this to our liking in the future without breaking changes, as Sam mentioned.